### PR TITLE
[Core] Checking number of active constraints on B&S

### PR DESCRIPTION
**Pull request by loumalouomega**
If the modelpart contained inactive constraints the assembling failed (potential unused dofs)

### CODE REVIEW

File: kratos/utilities/constraint_utilities.cpp [hunk 1]
```diff
@@ -22,12 +22,50 @@ namespace Kratos
...
+    std::size_t number_of_active_constraints = 0;
+
+    // The number of constraints
+    const int number_of_constraints = static_cast<int>(rModelPart.MasterSlaveConstraints().size());
+    const auto it_const_begin = rModelPart.MasterSlaveConstraints().begin();
+
+    // Setting to zero the slave dofs
+    #pragma omp parallel
+    {
+        #pragma omp for schedule(guided, 512) reduction(+:number_of_active_constraints)
```
Comment (philbucher) [inline kratos/utilities/constraint_utilities.cpp]:
```suggestion
        #pragma omp parallel for reduction(+:number_of_active_constraints)
```

also I think you don't need the outer `omp parallel`

**Review by philbucher — CHANGES_REQUESTED** [review]
I am not sure if this is right, bcs now if you set a constraint to ACTIVE/INACTIVE you have to rebuild the system!
This can easily be broken!!!
I would rather try to find the real problem

Also why should the constraints behave different from elems/conds?

File: kratos/utilities/constraint_utilities.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/constraint_utilities.cpp]: OK

**Comment by loumalouomega** [comment]
> I am not sure if this is right, bcs now if you set a constraint to ACTIVE/INACTIVE you have to rebuild the system!
> This can easily be broken!!!
When you activate/deactivate the constraints you need to rebuild the system indeed.
> I would rather try to find the real problem
I think this is the real problem, when you activate deactivate this may happen, and you need to prevent
> 
> Also why should the constraints behave different from elems/conds?

This can be applied for elements and conditiosn too, but is more common to have dynamic constraints that dynamic elem/cond

**Comment by philbucher** [comment]
> > I am not sure if this is right, bcs now if you set a constraint to ACTIVE/INACTIVE you have to rebuild the system!
> > This can easily be broken!!!
> > When you activate/deactivate the constraints you need to rebuild the system indeed.
> > I would rather try to find the real problem
> > I think this is the real problem, when you activate deactivate this may happen, and you need to prevent
> > Also why should the constraints behave different from elems/conds?
> 
> This can be applied for elements and conditiosn too, but is more common to have dynamic constraints that dynamic elem/cond

I understand, but then it is even more crucial to fix it porperly

What I don't get is why the assembly should fails. There may be zero-contributions, but I would say it should work in either case

**Comment by loumalouomega** [comment]
> > > I am not sure if this is right, bcs now if you set a constraint to ACTIVE/INACTIVE you have to rebuild the system!
> > > This can easily be broken!!!
> > > When you activate/deactivate the constraints you need to rebuild the system indeed.
> > > I would rather try to find the real problem
> > > I think this is the real problem, when you activate deactivate this may happen, and you need to prevent
> > > Also why should the constraints behave different from elems/conds?
> > 
> > 
> > This can be applied for elements and conditiosn too, but is more common to have dynamic constraints that dynamic elem/cond
> 
> I understand, but then it is even more crucial to fix it porperly
> 
> What I don't get is why the assembly should fails. There may be zero-contributions, but I would say it should work in either case

Imagine you have a constraint which works based on a "real" dof and an "auxiliar" dof. The real dofs are those are added by the elements and conditions, the auxiliar ones are those that are only added by the constraints. This is the case of a simplified contact implemented with MPC in a private branch, where there is an auxiliar dofs for each direction of the orthogonal base of the node. If you deactivate all the constraints in that case it will fail when trying to construct the LHS and RHS. This is because on the list of dofs these auxiliar dofs appear, but like the constraints are inactive, they actually don't exist.

May be consistent perhaps to check the activity on conditions and elements too. But what I am proposing is important to avoid certains implementations to fail in certain cases.

Another possible solution is to check that the equation id actually exists on the system actually assembled, but this could break more stuff that actually solve the problem.

**Comment by loumalouomega** [comment]
The discussion on this stopped, but we haven't reach any conclusion yet. Assigning to @KratosMultiphysics/implementation-committee

**Comment by philbucher** [comment]
@loumalouomega if you merge master into all of your stale PRs in the morning you basically block appveyor the entire day and no new PRs can be merged
Can you please cancel those builds and restart them in the evening when there is less activity?
Thanks a lot :)

**Comment by loumalouomega** [comment]
> 
> 
> @loumalouomega if you merge master into all of your stale PRs in the morning you basically block appveyor the entire day and no new PRs can be merged
> Can you please cancel those builds and restart them in the evening when there is less activity?
> Thanks a lot :)

Done

**Comment by philbucher** [comment]
thanks

**Comment by philbucher** [comment]
Hm I checked this again but I am not in favor of this change
The problem you describe sounds to me like sth different from our usual workflow that would maybe justify a derived B&S
What do you think @RiccardoRossi @adityaghantasala ?

**Comment by adityaghantasala** [comment]
@loumalouomega  

I activate and deactivate constraints in the ChimeraApp for fractional step. This has not caused any problems for me. 

When you say 

> The real dofs are those are added by the elements and conditions, the auxiliar ones are those that are only added by the constraints. 

I do not understand this. But as I understand, a constraint CANNOT create its own Dof(s). A constraint always should act on an existing constraint. As you say this might be the reason.

**Comment by loumalouomega** [comment]
> I do not understand this. But as I understand, a constraint CANNOT create its own Dof(s). A constraint always should act on an existing constraint. As you say this might be the reason.

We have some implementations doing that...

**Comment by adityaghantasala** [comment]
Can you point me to them ?

**Comment by adityaghantasala** [comment]
>  But as I understand, a constraint CANNOT create its own Dof(s). A constraint always should act on an existing constraint.

saying that now I think the following code is not necessary (may be for a different PR)

https://github.com/KratosMultiphysics/Kratos/blob/11b2be8e229d7d6f24eedd72fc4cfdb8a3bed735/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L572

**Comment by loumalouomega** [comment]
> 
> 
> Can you point me to them ?

Nope

**Comment by philbucher** [comment]
I checked this once again and I am still not in favor

even more without a usecase. Can you show one?

@KratosMultiphysics/technical-committee I guess we should discuss this

**Comment by loumalouomega** [comment]
> I checked this once again and I am still not in favor
> 
> even more without a usecase. Can you show one?
> 
> @KratosMultiphysics/technical-committee I guess we should discuss this

According to my first comment "If the modelpart contained inactive constraints the assembling failed (potential unused dofs)", I did this long time ago, maybe now it works, I need to do a test

**Comment by loumalouomega** [comment]
> I checked this once again and I am still not in favor
> 
> even more without a usecase. Can you show one?
> 
> @KratosMultiphysics/technical-committee I guess we should discuss this

Are we open to discuss this again?

**Comment by rubenzorrilla** [comment]
My two cents about this. I think that the constraints are expected to have the same behavior to that of the element and conditions. Hence,
1. Constraints are `ACTIVE` by default as elements and conditions are. Then, this change makes sense for me
```c++
bool constraint_is_active = true;
if (it_const->IsDefined(ACTIVE))
    constraint_is_active = it_const->Is(ACTIVE);
```
2. We only assemble their contribution if the constraint is `ACTIVE`.

In this regard, I'd add that, setting the condition to `ACTIVE` or `INACTIVE` doesn't require to reform the dofset but to apply a diagonal transformation matrix (I'm not 100% aware of the implementation details so please correct me if I'm wrong). I checked the current `master` branch and it seems that we're doing exactly this.

@loumalouomega maybe we should merge the `master` since I've the feeling that this issue is actually solved.

**Comment by loumalouomega** [comment]
> My two cents about this. I think that the constraints are expected to have the same behavior to that of the element and conditions. Hence,
> 1. Constraints are `ACTIVE` by default as elements and conditions are. Then, this change makes sense for me
> ```c++
> bool constraint_is_active = true;
> if (it_const->IsDefined(ACTIVE))
>     constraint_is_active = it_const->Is(ACTIVE);
> ```
> 2. We only assemble their contribution if the constraint is `ACTIVE`.
> 
> In this regard, I'd add that, setting the condition to `ACTIVE` or `INACTIVE` doesn't require to reform the dofset but to apply a diagonal transformation matrix (I'm not 100% aware of the implementation details so please correct me if I'm wrong). I checked the current `master` branch and it seems that we're doing exactly this.
> 
> @loumalouomega maybe we should merge the `master` since I've the feeling that this issue is actually solved. 

This is already merged with master. And there is a reason for this. The reason is that compute the constraints is expensive (require compute the connectivity of a T matrix), and then a matrix multiplication such TtKT which is not cheap for sparse systems. This can avoid a significative computation

**Comment by rubenzorrilla** [comment]
> > My two cents about this. I think that the constraints are expected to have the same behavior to that of the element and conditions. Hence,
> > 
> > 1. Constraints are `ACTIVE` by default as elements and conditions are. Then, this change makes sense for me
> > 
> > ```c++
> > bool constraint_is_active = true;
> > if (it_const->IsDefined(ACTIVE))
> >     constraint_is_active = it_const->Is(ACTIVE);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. We only assemble their contribution if the constraint is `ACTIVE`.
> > 
> > In this regard, I'd add that, setting the condition to `ACTIVE` or `INACTIVE` doesn't require to reform the dofset but to apply a diagonal transformation matrix (I'm not 100% aware of the implementation details so please correct me if I'm wrong). I checked the current `master` branch and it seems that we're doing exactly this.
> > @loumalouomega maybe we should merge the `master` since I've the feeling that this issue is actually solved.
> 
> This is already merged with master. And there is a reason for this. The reason is that compute the constraints is expensive (require compute the connectivity of a T matrix), and then a matrix multiplication such TtKT which is not cheap for sparse systems. This can avoid a significative computation

OK. Now I see your point. Thanks.

**Comment by loumalouomega** [comment]
> > > My two cents about this. I think that the constraints are expected to have the same behavior to that of the element and conditions. Hence,
> > > 
> > > 1. Constraints are `ACTIVE` by default as elements and conditions are. Then, this change makes sense for me
> > > 
> > > ```c++
> > > bool constraint_is_active = true;
> > > if (it_const->IsDefined(ACTIVE))
> > >     constraint_is_active = it_const->Is(ACTIVE);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 2. We only assemble their contribution if the constraint is `ACTIVE`.
> > > 
> > > In this regard, I'd add that, setting the condition to `ACTIVE` or `INACTIVE` doesn't require to reform the dofset but to apply a diagonal transformation matrix (I'm not 100% aware of the implementation details so please correct me if I'm wrong). I checked the current `master` branch and it seems that we're doing exactly this.
> > > @loumalouomega maybe we should merge the `master` since I've the feeling that this issue is actually solved.
> > 
> > This is already merged with master. And there is a reason for this. The reason is that compute the constraints is expensive (require compute the connectivity of a T matrix), and then a matrix multiplication such TtKT which is not cheap for sparse systems. This can avoid a significative computation
> 
> OK. Now I see your point. Thanks.

In any case ci is failing... :P

**Review by philbucher — COMMENTED** [review]
Can can review in detail again in the next days
Can you in the meantime please add a test to demonstrate the issue? I am still not sure I fully understand the problem

**Comment by loumalouomega** [comment]
> Can can review in detail again in the next days Can you in the meantime please add a test to demonstrate the issue? I am still not sure I fully understand the problem

A test is not needed, this helps to avoid the computation of the whole T operator and its corresponding multiplication, which is expensive in relatively big systems

**Comment by philbucher** [comment]
> > Can can review in detail again in the next days Can you in the meantime please add a test to demonstrate the issue? I am still not sure I fully understand the problem
> 
> A test is not needed, this helps to avoid the computation of the whole T operator and its corresponding multiplication, which is expensive in relatively big systems

In the initial comment you said that the assembly fails in some cases. Hence I am not clear if this is
a) an optimization of the computation time **OR**
b) a bugfix

can you clarify please?

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h [hunk 2]
```diff
@@ -219,7 +219,7 @@ class ResidualBasedEliminationBuilderAndSolverWithConstraints
 
     void SetUpSystem(ModelPart& rModelPart) override
     {
-        if(rModelPart.MasterSlaveConstraints().size() > 0)
+        if(ConstraintUtilities::NumberOfActiveConstraints(rModelPart) > 0)
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h:222]: how much is the performance penalty from looping?
I.e. if you have lots of constraints I can imagine this to become expensive too since is done many many times

**Review by philbucher — COMMENTED** [review]
the changes related to `ACTIVE` make sense to me

not sure about always checking whether some constraints are active is worth the effort though

**Review by philbucher — COMMENTED** [review]
ok so I now checked in detail why I did not like this, and my reasons have not changed (see my [initial comment](https://github.com/KratosMultiphysics/Kratos/pull/5223#pullrequestreview-261223086) from 2.5 years ago)
I checked and the proposed changes make the treatment of constraints DIFFERENT from the treatment of elements and conditions!
I.e. the dofs of elements and conditions are always assembled into the DofSet, regardless of whether they are active or not. The `ACTIVE` check is only done when it comes to the assembly of the contributions

However this PR changes the assembly of the constraints dofs! If a constraint is not active, then its dofs will not be part of the DofSet! Thus changing the `ACTIVE` state of constraints will require the system matrix to be rebuilt! (as some entries in the matrix might not be there due to missing dofs that we initially no added)

In summary my opinion has not changed, unless a solid usecase is provided I am not in favor of this change, sorry :/

**Comment by loumalouomega** [comment]
> ok so I now checked in detail why I did not like this, and my reasons have not changed (see my [initial comment](https://github.com/KratosMultiphysics/Kratos/pull/5223#pullrequestreview-261223086) from 2.5 years ago) I checked and the proposed changes make the treatment of constraints DIFFERENT from the treatment of elements and conditions! I.e. the dofs of elements and conditions are always assembled into the DofSet, regardless of whether they are active or not. The `ACTIVE` check is only done when it comes to the assembly of the contributions

Yes, but the MAIN DIFFERENCE is that the constraints require the computation of a sparse matrix, and it's corresponding multiplication with the LHS. This has a cost, even when there is only one constraint, and it is deactivated (the LHS stays the same therefore). So doing this will reduce computation when the constraints are deactivated, something that may happen if you consider the constraints for a contact problem for example.
> 
> However this PR changes the assembly of the constraints dofs! If a constraint is not active, then its dofs will not be part of the DofSet! Thus changing the `ACTIVE` state of constraints will require the system matrix to be rebuilt! (as some entries in the matrix might not be there due to missing dofs that we initially no added)

Usually, constraints relate dofs that are already in the system. But you are right, if those dofs are not in the connectivity system already, the computation will fail.

> 
> In summary my opinion has not changed, unless a solid usecase is provided I am not in favor of this change, sorry :/

If my argument for performance is not valid, I will close this PR.

**Comment by loumalouomega** [comment]
> > ok so I now checked in detail why I did not like this, and my reasons have not changed (see my [initial comment](https://github.com/KratosMultiphysics/Kratos/pull/5223#pullrequestreview-261223086) from 2.5 years ago) I checked and the proposed changes make the treatment of constraints DIFFERENT from the treatment of elements and conditions! I.e. the dofs of elements and conditions are always assembled into the DofSet, regardless of whether they are active or not. The `ACTIVE` check is only done when it comes to the assembly of the contributions
> 
> Yes, but the MAIN DIFFERENCE is that the constraints require the computation of a sparse matrix, and it's corresponding multiplication with the LHS. This has a cost, even when there is only one constraint, and it is deactivated (the LHS stays the same therefore). So doing this will reduce computation when the constraints are deactivated, something that may happen if you consider the constraints for a contact problem for example.
> 
> > However this PR changes the assembly of the constraints dofs! If a constraint is not active, then its dofs will not be part of the DofSet! Thus changing the `ACTIVE` state of constraints will require the system matrix to be rebuilt! (as some entries in the matrix might not be there due to missing dofs that we initially no added)
> 
> Usually, constraints relate dofs that are already in the system. But you are right, if those dofs are not in the connectivity system already, the computation will fail.
> 
> > In summary my opinion has not changed, unless a solid usecase is provided I am not in favor of this change, sorry :/
> 
> If my argument for performance is not valid, I will close this PR.

I will close this
